packages/timeline/src/sheets/build-nested-data.js

```diff
@@ -0,0 +1,166 @@
+      ]
+  OUTPUT:
+    [
+      [
+        [
+          groupFlag6,
+          [
+            [null, [record7, record8]],
+            [unitFlag9, [record10, record11]]
+          ]
+        ],
+      ],
+      [],
+      [],
+    ]
+*/
+
+/**
+ * Build the `section` to which the heading element belong.
+ * The elements and the subsections in the stacks will be aggregate into the section.
+ *
+ * @param {Object} headingElement
+ * @param {[][]} stacks
+ * @returns {[]} a section array like [headingElement, [...subsectionsOrElements]]
+ */
+function aggregateSection(headingElement, stacks) {
+  const newStacks = buildEmptyStacks(levelsCount)
+  const targetLevel = getElementHeadingLevel(headingElement.type)
+  for (let level = maxLevel; level > targetLevel; level--) {
+    const _levelStack = stacks[level]
+    if (_levelStack.length > 0) {
+      const _headingElement = targetLevel === level - 1 ? headingElement : null
+      const _section = [_headingElement, [..._levelStack]]
+      newStacks[level - 1].unshift(_section)
+    }
+  }
+  return newStacks[targetLevel][0]
+}
+
+/**
+  Transform flat elements into nested sections.
+  A `section` can be represent by an array with two items: `[headingElement, [...subsectionsOrElements]]`
+
+    INPUT:
+          [
+            groupFlag1,
+            unitFlag2,
+            record3,
+            groupFlag4, 
+            record5,
+            record6,
+            unitFlag7,
+            record8,
+            unitFlag9,
+            record10,
+            record11,
+          ]
+    OUTPUT:
+          [
+            [
+              groupFlag1,
+              [
+                [unitFlag2, [record3]]
+              ]
+            ],
+            [
+              groupFlag4,
+              [
+                [null, [record5, record6]],
+                [unitFlag7, [record8]],
+                [unitFla9, [record10, record11]]
+              ]
+            ]
+          ]
+ * @export
+ * @param {Object[]} elements
+ * @returns
+ */
+export function buildNestedData(elements) {
+  let _stack = buildEmptyStacks(levelsCount)
+  const elementsCount = elements.length
+  // loop from the end of list
+  for (let i = elementsCount - 1; i >= 0; i--) {
+    const element = elements[i]
+    const headingLevel = getElementHeadingLevel(element.type)
+    if (headingLevel === -1) {
+      // If the current element is not a heading element,
+      // insert it to the head of corresponding level.
+      // (it's only could be a `record` currently)
+      _stack[maxLevel].unshift(element)
+    } else {
+      // If the current element is a heading element,
+      // collect the elements above it and pack them into the subsection.
+      const newSection = aggregateSection(element, _stack)
+      _stack[headingLevel].unshift(newSection)
+      // Clear the packed elements
+      for (let level = maxLevel; level > headingLevel; level--) {
+        _stack[level] = []
+      }
+    }
+  }
+  return _stack[0]
```

Comment:
Since we return `_stack[0]`, do we need to inform users that 
spreadsheet rows needs the first row to be `group-flag` row?

Reply:
It should return the first level that `length>0`